### PR TITLE
[Fixes #499 #544] Allow importing from node_modules in lazy in-repo engines

### LINKED ISSUE #499: Cannot import anonymous AMD assets in lazy engine
Importing anonymous AMD assets inside lazy engines does not work:

`this.import('vendor/heatmap.js', { using: [ { transformation: 'amd', as: 'heatmap.js' } ] })`

results in the following error:

`TypeError: Cannot read property 'has' of undefined`

A debug session revealed that `this._customTransformsMap` is `undefined` when the import function is called from the engine addon. Replacing `this.import(...)` with `app.import(...)` resolves the problem, but results in the asset being hoisted into the host application.

### PR DESCRIPTION
**Pull request by villander**
this solves the conflicts and makes the #562  ready to merge

cc: @dgeb @stefanpenner

### CODE REVIEW

**Comment by villander** [comment]
@dgeb can we merge this? it's I'll save a lot of users with this issue ; )

**Comment by villander** [comment]
@stefanpenner can you take a look this, please?

**Comment by villander** [comment]
@rwjblue  can you take a look this, please?

**Comment by dbouwman** [comment]
@villander @rwjblue any movement on landing this? We are in the process of re-organizing a large app,  into a host app, + 5 engines. We already have a complex build process, and have been are running off @villander's branch - thus far it's working well. That said, if there is some issue w/ this solution we'd like to know sooner vs later. Thanks!

**Comment by XaserAcheron** [comment]
I wish I could say that this branch is working for me, but when I try it, I get stopped by what looks like some sort of bad path concatenation (note the weird "\C:" at the end):

```
  - broccoliBuilderErrorStack: Error: ENOENT: no such file or directory, mkdir 'C:\Users\Xaser\AppData\Local\Temp\broccoli-5220i4rbae6FTeLX\out-0925-funnel_funnel_custom_transform_amd\C:'
    at Object.fs.mkdirSync (fs.js:885:18)
    at Funnel.applyPatch [as _applyPatch] (C:\Code\GitHub\__redacted__\node_modules\ember-engines\node_modules\broccoli-funnel\index.js:382:10)
    at Funnel.<anonymous> (C:\Code\GitHub\__redacted__\node_modules\ember-engines\node_modules\broccoli-funnel\index.js:350:10)
    at Array.forEach (<anonymous>)
    at Funnel.processFilters (C:\Code\GitHub\__redacted__\node_modules\ember-engines\node_modules\broccoli-funnel\index.js:349:11)
    at Funnel.build (C:\Code\GitHub\__redacted__\node_modules\ember-engines\node_modules\broccoli-funnel\index.js:259:10)
    at resolve (C:\Code\GitHub\__redacted__\node_modules\broccoli\lib\wrappers\transform-node.js:29:35)
    at new Promise (<anonymous>)
    at TransformNodeWrapper.build (C:\Code\GitHub\__redacted__\node_modules\broccoli\lib\wrappers\transform-node.js:21:12)
    at Promise.resolve.then.then.then (C:\Code\GitHub\__redacted__\node_modules\broccoli\lib\builder.js:111:28)
  - code: [undefined]
  - codeFrame: ENOENT: no such file or directory, mkdir 'C:\Users\Xaser\AppData\Local\Temp\broccoli-5220i4rbae6FTeLX\out-0925-funnel_funnel_custom_transform_amd\C:'
  - errorMessage: ENOENT: no such file or directory, mkdir 'C:\Users\Xaser\AppData\Local\Temp\broccoli-5220i4rbae6FTeLX\out-0925-funnel_funnel_custom_transform_amd\C:'
        at Funnel (Funnel (custom transform: amd))
-~- created here: -~-
    at Funnel.Plugin (C:\Code\GitHub\__redacted__\node_modules\broccoli-plugin\index.js:7:31)
    at new Funnel (C:\Code\GitHub\__redacted__\node_modules\ember-engines\node_modules\broccoli-funnel\index.js:57:10)
    at Class.buildExternalTree (C:\Code\GitHub\__redacted__\node_modules\ember-engines\lib\engine-addon.js:71:25)
    at Class.<anonymous> (C:\Code\GitHub\__redacted__\node_modules\ember-engines\lib\utils\memoize.js:12:38)
    at Class.buildVendorCSSWithImports (C:\Code\GitHub\__redacted__\node_modules\ember-engines\lib\engine-addon.js:218:40)
    at Class.<anonymous> (C:\Code\GitHub\__redacted__\node_modules\ember-engines\lib\utils\memoize.js:12:38)
    at Class.compileLazyEngineStyles (C:\Code\GitHub\__redacted__\node_modules\ember-engines\lib\engine-addon.js:741:61)
    at Class.treeForPublic (C:\Code\GitHub\__redacted__\node_modules\ember-engines\lib\engine-addon.js:818:36)
    at Class._treeFor (C:\Code\GitHub\__redacted__\node_modules\ember-cli\lib\models\addon.js:684:33)
    at Class.treeFor (C:\Code\GitHub\__redacted__\node_modules\ember-engines\lib\engine-addon.js:1136:23)
-~- (end) -~-
  - errorType: Build Error
  - location:
    - column: [undefined]
    - file: [undefined]
    - line: [undefined]
    - treeDir: [undefined]
  - message: ENOENT: no such file or directory, mkdir 'C:\Users\Xaser\AppData\Local\Temp\broccoli-5220i4rbae6FTeLX\out-0925-funnel_funnel_custom_transform_amd\C:'
        at Funnel (Funnel (custom transform: amd))
-~- created here: -~-
    at Funnel.Plugin (C:\Code\GitHub\__redacted__\node_modules\broccoli-plugin\index.js:7:31)
    at new Funnel (C:\Code\GitHub\__redacted__\node_modules\ember-engines\node_modules\broccoli-funnel\index.js:57:10)
    at Class.buildExternalTree (C:\Code\GitHub\__redacted__\node_modules\ember-engines\lib\engine-addon.js:71:25)
    at Class.<anonymous> (C:\Code\GitHub\__redacted__\node_modules\ember-engines\lib\utils\memoize.js:12:38)
    at Class.buildVendorCSSWithImports (C:\Code\GitHub\__redacted__\node_modules\ember-engines\lib\engine-addon.js:218:40)
    at Class.<anonymous> (C:\Code\GitHub\__redacted__\node_modules\ember-engines\lib\utils\memoize.js:12:38)
    at Class.compileLazyEngineStyles (C:\Code\GitHub\__redacted__\node_modules\ember-engines\lib\engine-addon.js:741:61)
    at Class.treeForPublic (C:\Code\GitHub\__redacted__\node_modules\ember-engines\lib\engine-addon.js:818:36)
    at Class._treeFor (C:\Code\GitHub\__redacted__\node_modules\ember-cli\lib\models\addon.js:684:33)
    at Class.treeFor (C:\Code\GitHub\__redacted__\node_modules\ember-engines\lib\engine-addon.js:1136:23)
-~- (end) -~-
  - name: BuildError
  - nodeAnnotation: Funnel (custom transform: amd)
  - nodeName: Funnel
  - originalErrorMessage: ENOENT: no such file or directory, mkdir 'C:\Users\Xaser\AppData\Local\Temp\broccoli-5220i4rbae6FTeLX\out-0925-funnel_funnel_custom_transform_amd\C:'
  - stack: Error: ENOENT: no such file or directory, mkdir 'C:\Users\Xaser\AppData\Local\Temp\broccoli-5220i4rbae6FTeLX\out-0925-funnel_funnel_custom_transform_amd\C:'
    at Object.fs.mkdirSync (fs.js:885:18)
    at Funnel.applyPatch [as _applyPatch] (C:\Code\GitHub\__redacted__\node_modules\ember-engines\node_modules\broccoli-funnel\index.js:382:10)
    at Funnel.<anonymous> (C:\Code\GitHub\__redacted__\node_modules\ember-engines\node_modules\broccoli-funnel\index.js:350:10)
    at Array.forEach (<anonymous>)
    at Funnel.processFilters (C:\Code\GitHub\__redacted__\node_modules\ember-engines\node_modules\broccoli-funnel\index.js:349:11)
    at Funnel.build (C:\Code\GitHub\__redacted__\node_modules\ember-engines\node_modules\broccoli-funnel\index.js:259:10)
    at resolve (C:\Code\GitHub\__redacted__\node_modules\broccoli\lib\wrappers\transform-node.js:29:35)
    at new Promise (<anonymous>)
    at TransformNodeWrapper.build (C:\Code\GitHub\__redacted__\node_modules\broccoli\lib\wrappers\transform-node.js:21:12)
    at Promise.resolve.then.then.then (C:\Code\GitHub\__redacted__\node_modules\broccoli\lib\builder.js:111:28)

```

As a note, I'm on Windows, so I'm wondering if it's some sort of path join misbehavior.

I'm not even sure this is technically related to this issue, but I haven't run into this problem with amd transforms before, so it's all I can suspect.

**Comment by XaserAcheron** [comment]
Hmm, I managed to repro the above in a project without `ember-engines`, so scratch that -- it's indeed not related to this PR. Apologies for the red herring.

**Comment by villander** [comment]
no problem @XaserAcheron ; )

**Comment by XaserAcheron** [comment]
I got all my nonsense sorted out (t'was a bad export in the library I was trying to import), and I can confirm this PR indeed fixes the issue for me (was running into #544, specifically).

**Comment by thec0keman** [comment]
We've been working on a new engine for our Ember app, and this branch has resolved importing from node_modules on lazy engines.  (In our case it is not in-repo).

Is this ready for merge, or are there additional tasks to complete?

File: lib/utils/memoize.js [hunk 1]
```diff
@@ -3,13 +3,15 @@
 const calculateCacheKeyForTree = require('calculate-cache-key-for-tree');
 const CACHE = Object.create(null);
 
-module.exports = function memoize(func) {
-  return function _memoize() {
+function memoize(func) {
```
Comment (dgeb) [inline lib/utils/memoize.js]: Why switch this from a default export to a named export? Other util modules seem to use default exports.

File: package.json [hunk 2]
```diff
@@ -71,9 +71,10 @@
     "fs-extra": "^7.0.1",
     "lerna-changelog": "^0.8.2",
     "loader.js": "^4.7.0",
-    "mocha": "^6.0.0",
-    "prettier": "^1.16.4",
-    "walk-sync": "^1.1.3"
+    "mocha": "^5.2.0",
```
Comment (dgeb) [inline package.json]: What is the rationale for downgrading all three of these addons to earlier versions?

File: lib/utils/memoize.js (hunk 1 above)
Comment (villander) [reply lib/utils/memoize.js]: We need this export to mock this on tests with sinon. And we already have this export here - https://github.com/ember-engines/ember-engines/blob/master/lib/utils/append-compact-reexports-if-needed.js#L15

File: package.json (hunk 2 above)
Comment (villander) [reply package.json]: Fixed this, I undo these changes

File: lib/utils/memoize.js (hunk 1 above)
Comment (villander) [reply lib/utils/memoize.js]: @dgeb can we merge this?

**Comment by robertsteilberg** [comment]
Any way we can get this pushed through @rwjblue @dgeb? We are also depending on this branch and it's been sitting for quite some time.

**Comment by villander** [comment]
@robertsteilberg unfortunately we need of CI green here

**Comment by dbouwman** [comment]
@villander I've not worked with "core" code before, so I'm not exactly clear @ the issues noted @ Travis. It looks like it passes for 2.18, 3.4, but fails for the current ember release, beta and canary. 

The errors look to be the same across the failing versions- but, it's not at all clear to me what the underlying problem is, nor what a solution looks like.

If you can provide some guidance, I'd be happy to take a swing at getting this green.

**Comment by villander** [comment]
@dgeb CI is green now ; )

can we merge this

**Comment by dgeb** [comment]
Great! Thanks for everyone's patience :)

**Comment by villander** [comment]
@dgeb can you generate a new release for this, please?

**Comment by dgeb** [comment]
@villander - @rwjblue beat me to it :) 

https://github.com/ember-engines/ember-engines/releases/tag/v0.8.0

**Comment by villander** [comment]
thanks @rwjblue @dgeb

**Comment by dbouwman** [comment]
Thank you very much @villander, @dgeb and @rwjblue - much appreciated!

**Pull request merged**